crates/rpc/src/v03/method/get_events.rs

```diff
@@ -0,0 +1,893 @@
+        let keys: Vec<std::collections::HashSet<_>> = request
+            .keys
+            .into_iter()
+            .map(|keys| keys.into_iter().collect())
+            .collect();
+
+        let amount = request.chunk_size - events.events.len();
+
+        let skip = match count {
+            Some(count) => {
+                // This will not yield an underflow error, as when continuation_token is None,
+                // then the count is also always 0, since the last page can only be empty if there's
+                // only a single empty page
+                requested_offset.unwrap_or_default() - count
+            }
+            None => 0,
+        };
+
+        let is_last_page = append_pending_events(
+            &context.pending_data,
+            &mut events.events,
+            skip,
+            amount,
+            request.address,
+            keys,
+        )
+        .await;
+
+        events.continuation_token = next_continuation_token(
+            requested_offset.unwrap_or_default(),
+            request.chunk_size,
+            is_last_page,
+        );
+    }
+
+    check_continuation_token_validity(requested_offset, &events.events)?;
+
+    Ok(events)
+}
+
+// Maps `to_block` BlockId to a block number which can be used by the events query.
+//
+// This block id specifies the upper end of the range, so pending/latest/None means
+// there's no upper limit.
+fn map_to_block_to_number(
+    tx: &rusqlite::Transaction<'_>,
+    block: Option<BlockId>,
+) -> Result<Option<StarknetBlockNumber>, GetEventsError> {
+    use BlockId::*;
+
+    match block {
+        Some(Hash(hash)) => {
+            let number =
+                StarknetBlocksTable::get_number(tx, hash)?.ok_or(GetEventsError::BlockNotFound)?;
+
+            Ok(Some(number))
+        }
+        Some(Number(number)) => Ok(Some(number)),
+        Some(Pending) | Some(Latest) | None => Ok(None),
+    }
+}
+
+// Maps `from_block` BlockId to a block number which can be used by the events query.
+//
+// This block id specifies the lower end of the range, so pending/latest means
+// a lower limit here.
+fn map_from_block_to_number(
+    tx: &rusqlite::Transaction<'_>,
+    block: Option<BlockId>,
+) -> Result<Option<StarknetBlockNumber>, GetEventsError> {
+    use BlockId::*;
+
+    match block {
+        Some(Hash(hash)) => {
+            let number =
+                StarknetBlocksTable::get_number(tx, hash)?.ok_or(GetEventsError::BlockNotFound)?;
+
+            Ok(Some(number))
+        }
+        Some(Number(number)) => Ok(Some(number)),
+        Some(Pending) | Some(Latest) => {
+            let number =
+                StarknetBlocksTable::get_latest_number(tx)?.ok_or(GetEventsError::BlockNotFound)?;
+            Ok(Some(number))
+        }
+        None => Ok(None),
+    }
+}
+
+/// Append's pending events to `dst` based on the filter requirements and returns
+/// true if this was the last pending data i.e. `is_last_page`.
+async fn append_pending_events(
```

Comment:
nit: IMO it is counter-intuitive why `append_pending_events` returns `is_last_page` flag. 

If make it into something like `pull_pending_events` that would just return list of events (not append them to `dst`), then concept of `last_page` would not be necessary at all, as when last page was hit, the empty list would be returned, and the client code would know that the end of the list was reached.

Reply:
This and your other comment is perfectly valid. However,  this is basically just a copy-paste of the code we already had for v02, so I think these issues could be handled in a separate PR?